pkg/daemon/daemon.go

```diff
@@ -300,31 +402,48 @@ func (dn *Daemon) getHealth() error {
 //    desired machine state. if we aren't try updating again. if we are, update
 //    the current state annotation accordingly.
 func (dn *Daemon) CheckStateOnBoot() error {
-	// sanity check we're not already in a degraded state
-	if state, err := getNodeAnnotationExt(dn.kubeClient.CoreV1().Nodes(), dn.name, MachineConfigDaemonStateAnnotationKey, true); err != nil {
-		// try to set to degraded... because we failed to check if we're degraded
-		return dn.nodeWriter.SetUpdateDegradedIgnoreErr(err, dn.kubeClient.CoreV1().Nodes(), dn.name)
-	} else if state == MachineConfigDaemonStateDegraded {
-		// just sleep so that we don't clobber output of previous run which
-		// probably contains the real reason why we marked the node as degraded
-		// in the first place
+	stateAndConfigs, err := dn.getStateAndConfigs()
+	if err != nil {
+		return err
+	}
+	if stateAndConfigs.state == MachineConfigDaemonStateDegraded {
+		// We're already degraded.  Sleep so that we don't clobber
+		// output of previous run which probably contains the real
+		// reason why we marked the node as degraded in the first place
+		// (Though we should probably serialize the reason as an annotation)
 		glog.Info("Node is degraded; going to sleep")
 		select {}
 	}
 
-	// validate machine state
-	isDesired, dcAnnotation, err := dn.isDesiredMachineState()
+	if stateAndConfigs.currentConfig == stateAndConfigs.desiredConfig {
+		glog.Infof("In desired config %s", stateAndConfigs.currentConfig.GetName())
+		// Validate the on-disk state to catch any drift. We
+		// currently just do this on startup.
+		if isOnDiskValid := dn.validateOnDiskState(stateAndConfigs.desiredConfig); !isOnDiskValid {
+			return dn.nodeWriter.SetUpdateDegradedIgnoreErr(err, dn.kubeClient.CoreV1().Nodes(), dn.name)
+		} else {
+			glog.Info("Validated on-disk state")
+			// All good!
+			return nil
+		}
+	}
+
+	isPending, err := dn.isConfigCompletedPending(stateAndConfigs.desiredConfig.GetName())
 	if err != nil {
-		return dn.nodeWriter.SetUpdateDegradedIgnoreErr(err, dn.kubeClient.CoreV1().Nodes(), dn.name)
+		return err
 	}
 
-	if isDesired {
-		// we got the machine state we wanted. set the update complete!
-		if err := dn.completeUpdate(dcAnnotation); err != nil {
+	if isPending {
```

Comment:
I think here we should still re-validate the on-disk state before actually completing the update. We rebooted, and lots of services/scripts ran since we last applied the update. Also, in the context of automatic rollback, it's not as clear that a different boot ID implies "the rootfs we wanted to reboot into". In fact, it could be the same deployment, right? (E.g. if deployment finalization failed -- side note: we should probably enhance `pivot` to have a `--finalize` option or something so we do this upfront and fail early).

I think we should still also call `dn.checkOS()` in `validateOnDiskState()` for that reason. Checking the on-disk state is not that expensive when put in the context of how heavyweight a whole reboot process is already. So it's a mostly "free" operation in exchange for maintaining that crucial invariant that the `desiredConfig` annotation actually reflects reality.

Reply:
OK, makes sense.  Will do